curriculum/challenges/english/25-front-end-development/quiz-typescript/66f1ae758b77cfc3e4da6151.md

```diff
@@ -17,439 +17,454 @@ Answer all of the questions below correctly to pass the quiz.
 
 #### --answer--
 
-Placeholder answer
+Implementing a class
 
 ### --question--
 
 #### --text--
 
-Placeholder question
+What is the primary purpose of using generics in TypeScript?
 
 #### --distractors--
 
-Placeholder distractor 1
+To implement inheritance between classes.
 
 ---
 
-Placeholder distractor 2
+To define structure of an object.
 
 ---
 
-Placeholder distractor 3
+To create a new data type.
 
 #### --answer--
 
-Placeholder answer
+To write reusable code that can work with multiple types.
 
 ### --question--
 
 #### --text--
 
-Placeholder question
+Which of the following is a **proper example** of a generic function in TypeScript?
 
 #### --distractors--
 
-Placeholder distractor 1
+`function add(a: number, b: number): number { return a + b; }`
 
 ---
 
-Placeholder distractor 2
+`function multiply(a: number, b: number): { return a * b; }`
 
 ---
 
-Placeholder distractor 3
+`function getValue<T>(value: any): any { return value; }`
 
 #### --answer--
 
-Placeholder answer
+`function identity<T>(value: T): T { return value; }`
 
 ### --question--
 
 #### --text--
 
-Placeholder question
+What is type narrowing in TypeScript?
 
 #### --distractors--
 
-Placeholder distractor 1
+Reducing the size of compiled JavaScript
 
 ---
 
-Placeholder distractor 2
+Limiting the number of types in a union
 
 ---
 
-Placeholder distractor 3
+Removing unused types from a TypeScript project
 
 #### --answer--
 
-Placeholder answer
+To write code that narrows down a union type to a more specific type
 
 ### --question--
 
 #### --text--
 
-Placeholder question
+Which file is typically used to configure TypeScript compiler options?
 
 #### --distractors--
 
-Placeholder distractor 1
+`package.json`
 
 ---
 
-Placeholder distractor 2
+`package.config.json`
 
 ---
 
-Placeholder distractor 3
+`typescript.config.js`
 
 #### --answer--
 
-Placeholder answer
+`tsconfig.json`
 
 ### --question--
 
 #### --text--
 
-Placeholder question
+What does the `target` option in the `tsconfig.json` file specify?
 
 #### --distractors--
 
-Placeholder distractor 1
+The version of TypeScript to use during compilation.
 
 ---
 
-Placeholder distractor 2
+The target runtime environment (e.g., Node.js or browser) for the TypeScript code.
 
 ---
 
-Placeholder distractor 3
+The specific module system to use for imports and exports.
 
 #### --answer--
 
-Placeholder answer
+The ECMAScript version to which the TypeScript compiler will transpile the code.
 
 ### --question--
 
 #### --text--
 
-Placeholder question
+What does the `noImplicitReturns` option in the `tsconfig.json` file enforce?
 
 #### --distractors--
 
-Placeholder distractor 1
+It forces all functions to return a `number`.
 
 ---
 
-Placeholder distractor 2
+It requires that all functions return either `true` or `false`.
 
 ---
 
-Placeholder distractor 3
+It prevents using the `any` type in TypeScript code.
 
 #### --answer--
 
-Placeholder answer
+It warns if a function does not return a value.
```

Comment:
```suggestion
Checks all code paths in a function to ensure they return a value.
```
Might be a worthwhile distinction to add.

All functions implicitly return `undefined` and you can implicitly/explicitly have `undefined` as the return type, and Typescript will let the function "bottom out" without an explicit return.

---

Works
```
function isTrue(test: boolean) {
  if (test) {
    console.log("true");
  } else {
    return undefined;
  }
  console.log("Bottom");
}
isTrue(true);
```
Doesn't work
```
function isTrue(test: boolean) {
  if (test) {
    console.log("true");
  } else {
    return false;
  }
  console.log("Bottom");
}
isTrue(true);
```

Reply:
Got it. Learnt something new today.
Added the suggested changes